### PR TITLE
Labx 27

### PR DESCRIPTION
**Pull request by Polycarpik**
finished UserFactory. @lastperson please review

### CODE REVIEW

File: contracts/UserFactory.sol [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+            userLibrary.addRole(_user, _roles[i]);
+        }
+        return true;
+    }
+
+    function _setSkills(address _user, uint _areas, uint[] _categories, uint[] _skills) internal returns(bool){
+        if(_areas == 0){
+            return true;
+        }
+        userLibrary.setMany(_user, _areas, _categories, _skills);
```
Comment (lastperson) [inline contracts/UserFactory.sol]: return value is not checked, what if it fails?

File: test/userFactory.js [hunk 2]
```diff
@@ -0,0 +1,118 @@
+const EventsHistory = artifacts.require('./EventsHistory.sol');
+const UserLibraryMock = artifacts.require('./UserLibraryMock.sol');
+const UserFactory = artifacts.require('./UserFactory.sol');
+const ManagerMock = artifacts.require('./ManagerMock.sol');
```
Comment (lastperson) [inline test/userFactory.js]: Storage is not used in subject contracts, so ManagerMock is not needed too.

File: test/userFactory.js [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+  let userFactory;
+  let callCounter;
+  let storage;
+
+  let compareArrays = (arr1, arr2) => arr1.filter((item, i) => item !== arr2[i]).length === 0;
+
+  before('setup', () => {
+    return UserFactory.deployed()
+    .then(instance => userFactory = instance)
+    .then(() => Storage.deployed())
```
Comment (lastperson) [inline test/userFactory.js]: Storage not used.

File: test/userFactory.js [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+    .then(instance => storage = instance)
+    .then(() => ManagerMock.deployed())
```
Comment (lastperson) [inline test/userFactory.js]: ManagerMock not used.

File: test/userFactory.js [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+    .then(instance => eventsHistory = instance)
+    .then(() => userFactory.setupEventsHistory(eventsHistory.address))
+    .then(() => UserLibraryMock.deployed())
+    .then(instance => callCounter = instance)
+    .then(() => userFactory.setupUserLibrary(callCounter.address))
+    .then(reverter.snapshot);
+  });
+
+  it('should create users without roles and skills', () => {
+    return userFactory.createUserWithProxyAndRecovery(storage.address, [], "", [], [])
```
Comment (lastperson) [inline test/userFactory.js]: Single quotes are used throughout the project.

File: test/userFactory.js [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+    .then(result => {
+      assert.equal(result.logs.length, 1);
+      assert.equal(result.logs[0].event, 'UserCreated');
+      assert.equal(result.logs[0].args.roles.length, 0);
+      assert.equal(result.logs[0].args.areas.toString(), '0');
+      assert.equal(result.logs[0].args.categories.length, 0);
+      assert.equal(result.logs[0].args.skills.length, 0);
+      assert.equal(result.logs[0].args.proxy != null, true);
```
Comment (lastperson) [inline test/userFactory.js]: I think there is isNotNull assert available.

File: test/userFactory.js [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+    })
+    .then(() => callCounter.getCalls())
+    .then(array => {
+      assert.equal(array[0].toString(), '2');
+      assert.equal(array[1].toString(), '0');
+    });
+  });
+
+  it('should create users with skills', () => {
+    let areas = web3.toBigNumber(2).pow(2);
```
Comment (lastperson) [inline test/userFactory.js]: Why worry if this values are not checked anyway?

**Comment by Polycarpik** [comment]
Applied changes, @lastperson please review

**Pull request merged**